classes/Kohana/Log/Writer.php

```diff
@@ -67,24 +62,202 @@
+	 * 
+	 * @param string $timestamp_format
+	 * @return Log_Writer
+	 */
+	public function set_timestamp_format($timestamp_format)
+	{
+		$this->timestamp_format = (string) $timestamp_format;
+		
+		return $this;
+	}
+	
+	/**
+	 * Gets the log level used when tracing an exception stack
+	 * 
+	 * @return string the PSR-3 log level used for stack tracing
+	 */
+	public function get_strace_level()
+	{
+		return $this->strace_level;
+	}
+
+	/**
+	 * Sets the log level used when tracing an exception stack
+	 * 
+	 * @param mixed $level
+	 * @throws InvalidArgumentException
+	 * @return Log_Writer
+	 */
+	public function set_strace_level($level)
+	{
+		$this->strace_level = Log::to_psr_level($level);
+		
+		return $this;
+	}
+	
+	/**
+	 * Gets the default string format of the log entry
+	 * 
+	 * @return string
+	 */
+	public function get_format()
+	{
+		return $this->format;
+	}
+
+	/**
+	 * Sets the default string format of the log entry
+	 * 
+	 * @param string $format
+	 * @return Log_Writer
+	 */
+	public function set_format($format)
+	{
+		$this->format = (string) $format;
+		
+		return $this;
+	}
+	
+	/**
+	 * Gets the timezone of this Log_Writer
+	 * 
+	 * @return string
+	 */
+	public function get_timezone()
+	{
+		return $this->timezone;
+	}
+
+	/**
+	 * Sets the timezone of this Log_Writer
+	 * 
+	 * @param string $timezone
+	 * @throws InvalidArgumentException
+	 * @return Log_Writer
+	 */
+	public function set_timezone($timezone)
+	{
+		if ( ! in_array($timezone, DateTimeZone::listIdentifiers()))
+		{
+			throw new InvalidArgumentException(
+				'Argument 1 of set_timezone must be a valid timezone.'
+			);
+		}
+
+		$this->timezone = $timezone;
+
+		return $this;
+	}
+
+	/**
+	 * Gets the log levels that this writer accepts to write
+	 *
+	 * @return array
+	 */
+	public function get_levels()
+	{
+		return $this->levels;
+	}
+
+	/**
+	 * Sets the log levels that this writer accepts to write
+	 *
+	 * @param array $levels
+	 * @throws InvalidArgumentException
+	 * @return Log_Writer
+	 */
+	public function set_levels(array $levels)
+	{
+		$this->levels = array_map('Log::to_psr_level', $levels);
+
+		return $this;
+	}
+
+	/**
+	 * Sets the log levels' range that this writer accepts to write
+	 *
+	 * @param array $levels
+	 * @throws InvalidArgumentException
+	 * @return Log_Writer
+	 */
+	public function set_levels_range($min_level, $max_level)
+	{
+		$min_level = Log::to_int_level($min_level);
+		$max_level = Log::to_int_level($max_level);
+
+		if ( ! $min_level > $max_level)
+		{
+			throw InvalidArgumentException('maximum level should be greater than minimum level');
+		}
+
+		$this->levels = array_map('Log::to_psr_level', range($min_level, $max_level));
+
+		return $this;
+	}
+
+	/**
+	 * Filter messages according to the levels accepted by this writer
+	 *
+	 * @param array $messages
+	 * @return array Filtered messages
+	 */
+	public function filter(array $messages)
+	{
+		if (empty($this->levels))
```

Comment:
I wouldn't expect `NULL` to make it log everything either :) I'd prefer if it was explicit - default $levels to contain all levels.

If it's just that you want to skip filtering for performance if it's not required (?) then if $levels was a map of level => true/false you could do `if ($this->levels === array_filter($this->levels))` to know whether any have been excluded.